### ISSUE TITLE
pkg/dds type error in .catchError callback leading to runtime errors

### ISSUE THREAD

**Issue by christopherfujino** [dart-lang/sdk#50802]
The analyzer ignore [here](https://github.com/dart-lang/sdk/blob/3675fd25f4d0ef8038b7c7db316d7240a831f67a/pkg/dds/lib/src/dap/adapters/dart.dart#L1981) is obscuring a type error that is leading to runtime `ArgumentError` exceptions in the flutter CLI tool: https://github.com/flutter/flutter/issues/114031. This is currently the top crasher in the current Flutter beta release (3.7.0-1.2.pre).

**Comment by christopherfujino** [comment]
I tried to fix this by changing the return type of `ProtocolConverter.convertVmInstanceRefToDisplayString` from `Future<String>` -> `Future<String?>` (which I believe would have been the correct type when this code was initially migrated to null-safety) this broke several other call-sites, so I wasn't confident this was the right change.

**Comment by christopherfujino** [comment]
cc @bkonyi & @DanTup

**Comment by lrhn** [comment]
A better migration of the `catchError` could be:
```dart
   .then<String?>((s) => s, onError: (Object e, _) {
     logger?.call('$e');
     return null;
   });
```

**Comment by DanTup** [comment]
Yeah, I think we can leave `convertVmInstanceRefToDisplayString` as-is, and just allow the `null` in this call in case of an error.

@christopherfujino I can take a look in the morning if you don't beat me to it.

I'm curious about the exception leading to this though.. this `catch` is intended to handle some edge cases (such as a hot reload or collection just before we called `toString()` for an object). If it's happening a lot, there may be other issues that need investigating (I don't know if we have a good way to get info on those exceptions without either a repro or being able to log exceptions without them being unhandled and terminating the DAP server).

**Comment by DanTup** [comment]
There are two places that call this code (`getFullString`)

1. the `dart:developer` `log()` handler
2. the `pauseException` handler

Both of these may try to build strings to show the user, and that might require calling the VM Service. Depending on timing, it's possible the instances we're trying to show have been collected (or some other error occurs perhaps due to a hot restart or something).

I can't reproduce this using a real Flutter app, and I also can't get a test around this to simulate the issue without some refactoring. I feel like trying to get a fix into Flutter now would be quite risky (since this comes via `package:dds`, we'd need to branch a version of DDS from the commit that was released into Flutter beta, cherry-pick a fix there, then roll that version of DDS into Flutter).

So, my feeling is that maybe the best thing to do is hang fire on enabling the SDK DAP for Flutter VS Code users until the release after this one, and we can land a fix for this in master (with any necessary refactoring to ensure these cases are better tested).

Alternatively we could land @lrhn's suggestion above, but I don't know how we can get a test around it. It would also require some `package:dds` branching described above though.

Note: currently the SDK DAPs will _not_ automatically be activated for users on stable, this is happening for `beta` because we enable the SDK DAPs for all pre-release versions of Flutter (to get more testing and find issues like this). For peace of mind, I could disable that in the upcoming VS Code release so the `beta` crashes will go away, and we can re-enable that only for pre-release versions that have a fix for this once it's made.

Thoughts @christopherfujino @jacob314 ?

**Comment by DanTup** [comment]
> my feeling is that maybe the best thing to do is hang fire on enabling the SDK DAP for Flutter VS Code users until the release after this one

I decided to go ahead with this for now. It turned out I mis-remembered and we would have still enabled the SDK DAPs for 10% of users when there's a Flutter stable `>= 3.4.0-33`. Since there can be a delay in people updating their extensions I think it's better to disable this now, and we can always revert it if we get a fix (a delay in people getting the revert is less of an issue).

In addition to this issue, we didn't manage to get a fix for the Inspector navigating in the IDE, so I was already unsure about starting to roll people onto the SDK DAP for this version.

I had a small Dart-Code release prepared for next week but I've included this fix and published it today so hopefully a drop in these crashes will show up sooner (@christopherfujino please let me know if you don't see this).

To summarise:

- this issue only affects users of the new SDK DAP (I suspect it may also only really affect Flutter, as hot reload/restart may be involved)
- only VS Code users on Flutter pre-release SDKs are opted in to those DAPs today, but 10% of stable users would have gotten it if this version went stable
- `v3.56.0` of the VS Code extension (published today) will bump that minimum version number to 9.9.9 so no users will be opted in to the SDK DAP automatically (they can still opt-in manually)
- once a fix is available, the minimum version will be set accordingly (if a fix is CP'd, it can be set to include this beta, and otherwise a future version)

**Comment by christopherfujino** [comment]
> > my feeling is that maybe the best thing to do is hang fire on enabling the SDK DAP for Flutter VS Code users until the release after this one
> 
> I decided to go ahead with this for now. It turned out I mis-remembered and we would have still enabled the SDK DAPs for 10% of users when there's a Flutter stable `>= 3.4.0-33`. Since there can be a delay in people updating their extensions I think it's better to disable this now, and we can always revert it if we get a fix (a delay in people getting the revert is less of an issue).
> 
> In addition to this issue, we didn't manage to get a fix for the Inspector navigating in the IDE, so I was already unsure about starting to roll people onto the SDK DAP for this version.
> 
> I had a small Dart-Code release prepared for next week but I've included this fix and published it today so hopefully a drop in these crashes will show up sooner (@christopherfujino please let me know if you don't see this).
> 
> To summarise:
> 
> * this issue only affects users of the new SDK DAP (I suspect it may also only really affect Flutter, as hot reload/restart may be involved)
> * only VS Code users on Flutter pre-release SDKs are opted in to those DAPs today, but 10% of stable users would have gotten it if this version went stable
> * `v3.56.0` of the VS Code extension (published today) will bump that minimum version number to 9.9.9 so no users will be opted in to the SDK DAP automatically (they can still opt-in manually)
> * once a fix is available, the minimum version will be set accordingly (~if a fix is CP'd, it can be set to include this beta~ (see edit below), and otherwise a future version)
> 
> **Edit:** We're going to hold off enabling this until the Inspector issue is resolved too, so there's no value in trying to CP this - it will be disabled for everyone for a little longer.

Thanks Danny! I do see some reports of this on the latest stable release, but fewer than previously, and it came out a few days before this comment. Also, since the stacktrace is so uninformative, it's possible we're getting this from other callsites as well, unrelated to DDS. I will continue to monitor the numbers.

**Comment by DanTup** [comment]
It's possible some people haven't updated their extensions yet (some people disable auto-update, and others leave VS Code open for long periods, which prevents the new version from being started). Hopefully it'll continue to reduce (at least where it's coming from DDS).

Let me know if you continue to see concerning numbers though, and I'll gate any further enabling of the SDK daps on a version with a fix for this once done.
